File: app/javascript/definitions/PaymentIntent.ts
export interface PaymentPlan {
  count: number
  interval: string
  type: string
}

export interface Installments {
  available_plans: PaymentPlan[]
  enabled: boolean
  plan: PaymentPlan
}

export interface Card {
  installments: Installments
  network?: any
  request_three_d_secure: string
}

export interface PaymentMethodOptions {
  card: Card
}


export interface PaymentIntent {
  id: number
  payment_method_id: number
  user_id: number
  amount: number
  amount_received: number
  idempotency_key: string
  status: string
  payment_method_options: PaymentMethodOptions
  stripe_id: string
  created_at: Date
  updated_at: Date
}

File: app/models/payment_method/payment_intent.rb
class PaymentMethod::PaymentIntent
  attr_reader :user, :payment_method, :amount, :payment_intent

  def initialize(user, payment_method, amount)
    @user = user
    @amount = amount
    @payment_method = payment_method
  end

  def create
    @payment_intent = Stripe::PaymentIntent.create(
      amount: amount,
      currency: 'mxn',
      customer: user.stripe_id,
      payment_method: payment_method.stripe_id,
      payment_method_options: {
        card: {
          installments: {
            enabled: true
          }
        }
      }
    )
  end

  def id
    @payment_intent.id
  end

  def available_plans
    @payment_intent.payment_method_options.card.installments.available_plans
  end
end
